mirage/awa_mirage.ml

```diff
@@ -148,18 +148,24 @@ module Make (F : Mirage_flow.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) =
     FLOW.close t.flow
 
   let shutdown t mode =
-    (match t.state with
-     | `Active ssh | `Read_closed ssh | `Write_closed ssh ->
-       let ssh, msg =
-         match t.state, mode with
-         | (`Active ssh | `Read_closed ssh), `write -> Awa.Client.eof ssh
-         | _, `read_write -> Awa.Client.close ssh
-         | _ -> ssh, None
-       in
-       t.state <- inject_state ssh (half_close t.state mode);
-       writev_flow t (Option.to_list msg) >|= ignore
-     | `Error _ | `Closed -> Lwt.return_unit) >>= fun () ->
-    FLOW.shutdown t.flow mode
+    match t.state with
+    | `Active ssh | `Read_closed ssh | `Write_closed ssh ->
+      let ssh, msg =
+        match t.state, mode with
+        | (`Active ssh | `Read_closed ssh), `write -> Awa.Client.eof ssh
+        | _, `read_write -> Awa.Client.close ssh
+        | _ -> ssh, None
+      in
+      t.state <- inject_state ssh (half_close t.state mode);
+      writev_flow t (Option.to_list msg) >>= fun _ ->
+      (* we don't [FLOW.shutdown _ mode] because we still need to read/write
+         channel_eof/channel_close unless both directions are closed *)
+      (match t.state with
+       | `Closed ->
+         (* also close on error?! *)
```

Comment:
the answer is no, we don't want to close on error, since that has awkward semantics in respect to half-closes.

Reply:
```suggestion
```

Reply:
removed comment in 12040f4